sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalentExpressions.scala

```diff
@@ -106,79 +106,92 @@ class GpuEquivalentExpressions {
+      if (c.children.exists(hasSideEffects)) {
+        c.children.head :: Nil
+      } else {
+        c.children
+      }
     case other => other.children
   }
 
   // For some special expressions we cannot just recurse into all of its children, but we can
   // recursively add the common expressions shared between all of its children.
   private def commonChildrenToRecurse(expr: Expression): Seq[Seq[Expression]] = expr match {
     case _: CodegenFallback => Nil
-    case i: If => Seq(Seq(i.trueValue, i.falseValue))
-    case i: GpuIf => Seq(Seq(i.trueExpr, i.falseExpr))
-    case c: CaseWhen =>
-      // We look at subexpressions in conditions and values of `CaseWhen` separately. It is
-      // because a subexpression in conditions will be run no matter which condition is matched
-      // if it is shared among conditions, but it doesn't need to be shared in values. Similarly,
-      // a subexpression among values doesn't need to be in conditions because no matter which
-      // condition is true, it will be evaluated.
-      val conditions = if (c.branches.length > 1) {
-        c.branches.map(_._1)
-      } else {
-        // If there is only one branch, the first condition is already covered by
-        // `childrenToRecurse` and we should exclude it here.
-        Nil
-      }
-      // For an expression to be in all branch values of a CaseWhen statement, it must also be in
-      // the elseValue.
-      val values = if (c.elseValue.nonEmpty) {
-        c.branches.map(_._2) ++ c.elseValue
+    case i: GpuIf =>
+      if (hasSideEffects(i.trueExpr) || hasSideEffects(i.falseExpr)) {
+        Seq(Seq(i.trueExpr, i.falseExpr))
       } else {
         Nil
       }
-      Seq(conditions, values)
     case c: GpuCaseWhen =>
-      // We look at subexpressions in conditions and values of `CaseWhen` separately. It is
-      // because a subexpression in conditions will be run no matter which condition is matched
-      // if it is shared among conditions, but it doesn't need to be shared in values. Similarly,
-      // a subexpression among values doesn't need to be in conditions because no matter which
-      // condition is true, it will be evaluated.
-      val conditions = if (c.branches.length > 1) {
-        c.branches.map(_._1)
+      if (c.hasSideEffects) {
+        // We look at subexpressions in conditions and values of `CaseWhen` separately. It is
+        // because a subexpression in conditions will be run no matter which condition is matched
+        // if it is shared among conditions, but it doesn't need to be shared in values.
+        // Similarly,
+        // a subexpression among values doesn't need to be in conditions because no matter which
+        // condition is true, it will be evaluated.
+        val conditions = if (c.branches.length > 1) {
+          // We don't strip off the first condition expression because it is the only one
+          // that is guaranteed to be executed.
+          c.branches.map(_._1)
```

Comment:
The case above only looks at the very first expression of the first branch, but this grabs the predicate expression of all branches.  These are not guaranteed to be executed when there's side effects, as it will short-circuit sometimes in such a way that some of the branch predicates will never be evaluated.  Sort of like the GpuIf case, don't we want to be examining whether the branch predicate expressions have side effects?

Comment mentions "the first condition expression" but this grabs _all_ conditional expressions which is confusing.